src/sql/workbench/services/executionPlan/common/executionPlanService.ts

```diff
@@ -27,19 +28,43 @@ export class ExecutionPlanService implements IExecutionPlanService {
 		this._providerRegisterEvent = this._onProviderRegister.event;
 	}
 
-	private async ensureCapabilitiesRegistered(): Promise<void> {
-		let providers = Object.keys(this._capabilitiesService.providers);
-		if (!providers) {
-			await new Promise<void>(resolve => {
-				this._capabilitiesService.onCapabilitiesRegistered(e => {
-					resolve();
-				});
-			});
+	/**
+	 * This ensures that the capabilities service has registered the providers to handle execution plan requests.
+	 * @param providerId Optional provider id to wait for.
+	 */
+	private async ensureCapabilitiesRegistered(providerId?: string): Promise<void> {
+		let providers: string[] = Object.keys(this._capabilitiesService.providers);
+
+		// If the provider is already registered, return.
+		if (providerId && providers?.includes(providerId) || (!providerId && providers?.length > 0)) {
```

Comment:
Is a falsy providerId even valid? Are there cases where we're passing in an empty string or something?

Reply:
There are 2 cases for using execution plans.
1. Getting execution plan when we run queries. In this case we already have the provider id.
1. When we a open a file that is registered to be handled by execution plan editor. In this case, we do not know which provider within the execution plan service will be able to handle that file.

Reply:
But surely the service can figure that out? Otherwise how do you know which provider to send it to?

Reply:
The service can figure out only after the providers are registered. As during registration, they will provide the file types they support.

Reply:
So shouldn't it be : 

1. Check if a provider is registered that can handle it
2. If not then set up promise with event on capabilities registered to watch for provider being registered
3. For every provider registered see if it can handle the EP. If it can then complete the promise, if not then ignore
4. After certain timeout (15-30sec?) reject with "No provider registered" error

Reply:
I agree. I had changed the code because I misunderstood how capabilities providers worked and the if statement I was using in the code was wrong. The provider keys will never be null. Instead it will be an empty array. Now I waiting for the right provider to be registered.